### PR TITLE
3620 validate location barcode ids

### LINKED ISSUE #3620: Integration: Catch missing Barcode and Locations
In some datafiles there are missing barcodes and locations. Catch these so that these records can be integrated.

This has been observed fro:
- [x] Stock_take_lines
- [x] item_line
- [ ] trans_line

### PR DESCRIPTION
**Pull request by clemens-msupply**
Fixes #3620

# 👩🏻‍💻 What does this PR do? 

 

Validate location/barcode ids for invoice_line, stock_line and stocktake_line

I only fixed cases that I noticed in real datafiles, there might be other location_id references as well, could do them later...

### CODE REVIEW

File: server/service/src/sync/translations/invoice.rs [hunk 1]
```diff
@@ -260,7 +260,7 @@ impl SyncTranslation for InvoiceTranslation {
             .map(|store_row| store_row.id);
 
         let invoice_type = invoice_type(&data, &name).ok_or(anyhow::Error::msg(format!(
-            "Unsupported invoice type: {:?} for {:?} mode",
+            "Unsupported invoice type: {:?} for (mode: {:?})",
```
Comment (clemens-msupply) [inline server/service/src/sync/translations/invoice.rs:263]: Site tracked, got confused by this message...

File: server/service/src/sync/translations/stock_line.rs [hunk 2]
```diff
@@ -68,27 +71,45 @@ impl SyncTranslation for StockLineTranslation {
...
+            location_ID,
+            pack_size,
+            available,
+            quantity,
+            cost_price,
+            sell_price,
+            note,
+            supplier_id,
+            barcode_id,
+        } = serde_json::from_str::<LegacyStockLineRow>(&sync_record.data)?;
```
Comment (clemens-msupply) [inline server/service/src/sync/translations/stock_line.rs:93]: Some refactor along the way...

File: server/service/src/sync/translations/invoice.rs (hunk 1 above)
Comment (mark-prins) [reply server/service/src/sync/translations/invoice.rs:263]: not sure about the brackets. I would either remove or have them around the `for` as well 🤷 

```suggestion
            "Unsupported invoice type: {:?} for mode: {:?}",
```
or 

```suggestion
            "Unsupported invoice type: {:?} (for mode: {:?})",
```

**Review by roxy-dao — APPROVED** [review]
LGTM. Thanks Clemens

File: server/service/src/sync/translations/utils.rs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+            .find_one_by_id(&id)?
+            .map(|it| it.id)
+    } else {
+        None
+    };
+    Ok(barcode_id)
+}
+
+/// Some datafile contain links to non-existing location references.
+/// Check if the entry exist and if not return None.
```
Comment (mark-prins) [inline server/service/src/sync/translations/utils.rs:20]: :nit
```suggestion
/// Some datafiles contain links to non-existing location references.
/// Check if the entry exists and if not return None.
```

File: server/service/src/sync/translations/utils.rs [hunk 4]
```diff
@@ -0,0 +1,33 @@
+use repository::{BarcodeRowRepository, LocationRowRepository, RepositoryError, StorageConnection};
+
+/// Some datafile contain links to non-existing barcode references.
+/// Check if the entry exist and if not return None.
```
Comment (mark-prins) [inline server/service/src/sync/translations/utils.rs:4]: :nit

```suggestion
/// Some datafiles contain links to non-existing barcode references.
/// Check if the entry exists and if not return None.
```

**Review by mark-prins — COMMENTED** [review]
Oh?! Roxy approved while I was looking through 🐌 

Changes look good to me too!

**Pull request merged**